### PR TITLE
A DigestCalculator for the Alpakka Project

### PR DESCRIPTION
**Pull request by dnvriend**

### CODE REVIEW

File: digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import scala.concurrent.Future
+import scala.util.Success
+
+/**
+ * The DigestCalculator transforms/digests a stream of akka.util.ByteString to a
+ * DigestResult according to a given Algorithm
+ */
+object DigestCalculator {
+
+  def flow(algorithm: Algorithm): Flow[ByteString, DigestResult, NotUsed] =
```
Comment (patriknw) [inline digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala]: we decided to use scaladsl/javadsl packages, as in https://github.com/akka/alpakka/pull/19

File: digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = {
+        val chunk = grab(in)
+        digest.update(chunk.toArray)
+        pull(in)
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        val messageDigest: Array[Byte] = digest.digest()
+        emit(out, DigestResult(ByteString(messageDigest), digestToHexString(messageDigest), Success(Done)))
```
Comment (patriknw) [inline digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala]: when will it be Failure?

Comment (dnvriend) [inline digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala]: Good question! I believe it cannot happen at all. All other exceptions will cause the stream to fail. The interface I copied of `akka.stream.IOResult`.  Maybe it needs another interface?

File: digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala (hunk 1 above)
Comment (dnvriend) [inline digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala]: Java... alright, let me look into that!

File: digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala (hunk 2 above)
Comment (patriknw) [inline digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala]: ok, then you can remove the `status` from the `DigestResult`

File: digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+private[digest] class DigestCalculator(algorithm: Algorithm) extends GraphStage[FlowShape[ByteString, DigestResult]] {
+  val in: Inlet[ByteString] = Inlet("DigestCalculator.in")
+  val out: Outlet[DigestResult] = Outlet("DigestCalculator.out")
+  override val shape: FlowShape[ByteString, DigestResult] = FlowShape.of(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    val digest = java.security.MessageDigest.getInstance(algorithm.toString)
+
+    def digestToHexString(messageDigest: Array[Byte]): String =
```
Comment (patriknw) [inline digest/src/main/scala/akka/stream/alpakka/digest/DigestCalculator.scala]: perhaps this should be a `def` in DigestResult instead, so you don't have to pay for the hex string unless you need it

**Comment by drewhk** [comment]
LGTM

**Comment by patriknw** [comment]
@dnvriend Do you have any plans of completing this, or shall we close this PR for now?

**Pull request closed without merging**